Add Azure as provider

zechenbit opened the pull request:
## Description
Added Azure as provider. https://github.com/inlets/inlets-operator/issues/47

## How Has This Been Tested?
### Create ###
```
inletsctl create --provider=azure --subscription-id=4d68b80e-642c-4a95-8629-c9abe13d4cdb --region=eastus --access-token-file=/Users/chen/client_credentials.json #gosetup
Using provider: azure
Requesting host: friendly-blackburn1 in eastus, from azure
2020/02/02 23:20:18 Provisioning host with Azure
2020/02/02 23:20:24 Group created inlets-friendly-blackburn1
2020/02/02 23:20:29 Waiting for deployment completion...
2020/02/02 23:23:37 SSH User: inletsuser / 'xt7mbb8oVuV0O9ak'
Host: inlets-friendly-blackburn1, status: active
[1/500] Host: inlets-friendly-blackburn1, status: active
Inlets OSS exit-node summary:
  IP: 40.76.203.69
  Auth-token: FnDQF72If8PjT5cwBVpcgXAwKfcFUFry9XbfNCVFt5jf2sIK7zCEQIU1C8WJQYbU

Command:
  export UPSTREAM=http://127.0.0.1:8000
  inlets client --remote "ws://40.76.203.69:8080" \
        --token "FnDQF72If8PjT5cwBVpcgXAwKfcFUFry9XbfNCVFt5jf2sIK7zCEQIU1C8WJQYbU" \
        --upstream $UPSTREAM

To Delete:
        inletsctl delete --provider azure --id "inlets-friendly-blackburn1"
```
### Delete ###
```
inletsctl delete --provider=azure --id inlets-clever-volhard8 --subscription-id=4d68b80e-642c-4a95-8629-c9abe13d4cdb --region=eastasia --access-token-file=/Users/chen/client_credentials.json #gosetup
Using provider: azure
Deleting host: inlets-clever-volhard8 from azure
2020/02/02 23:13:19 Waiting for deletion completion (~10 mins)...
2020/02/02 23:25:06 Done deleting resources...
```


## How are existing users impacted? What migration steps/scripts do we need?
No Impact to existing users as this is a new feature.


## Checklist:

I have:

- [x] updated the documentation and/or roadmap (if required)
- [x] read the [CONTRIBUTION](https://github.com/inlets/inlets/blob/master/CONTRIBUTING.md) guide
- [x] signed-off my commits with `git commit -s`
- [x] added unit tests

alexellis commented on cmd/create.go:
```diff
@@ -34,6 +35,7 @@ func init() {
 	createCmd.Flags().String("secret-key-file", "", "Read this file for the access token for your cloud (Scaleway, EC2)")
 	createCmd.Flags().String("organisation-id", "", "Organisation ID (Scaleway)")
 	createCmd.Flags().String("project-id", "", "Project ID (Packet.com, Google Compute Engine)")
+	createCmd.Flags().String("subscription-id", "", "Subscription ID (Azure)")
```
I wonder if we should use project ID?

zechenbit replied:
Azure subscription ID is similar to GCE's project ID. However the term `project ID` is unclear to the Azure user what it is. We would need to explain it somewhere if we use `project ID`. So I prefer to use `subscription ID`.

alexellis replied:
I'm still in two minds about this, I think `project-id` might be easier to manage, but you have a good point too.

alexellis commented on cmd/create.go:
```diff
@@ -132,7 +134,17 @@ func runCreate(cmd *cobra.Command, _ []string) error {
+	if provider == "azure" {
+		subscriptionID, _ = cmd.Flags().GetString("subscription-id")
+		authFile, _ := cmd.Flags().GetString("access-token-file")
+		err = os.Setenv("AZURE_AUTH_LOCATION", authFile)
```
Can you rather pass the variable to the SDK than affect the OS environment variables?

zechenbit replied:
I had a look at the code, it does not seem to support passing as a variable: https://github.com/Azure/go-autorest/blob/master/autorest/azure/auth/auth.go#L295

Also their doc is using env for all the example, https://github.com/Azure/azure-sdk-for-go#authentication

So I'm not sure if passing as go variable instead of env is doable without changing the SDK. I will reach out to Azure GO SDK team.

alexellis replied:
Can you do this differently then? Can you automatically set the value inside the provisioner, so that it's fully encapsulated for all users / calls.

alexellis commented on cmd/create.go:
```diff
@@ -313,7 +326,23 @@ func createHost(provider, name, region, zone, projectID, userData, inletsPort st
+		return &provision.BasicHost{
+			Name:     name,
+			OS:       "Additional.imageOffer",
+			Plan:     "Standard_B1ls",
```
Is this the cheapest option with an IP?

zechenbit replied:
Yes it is the cheapest plan, with 1 core, 0.5 GiB RAM, ~$4 / month (depending on region). see https://azureprice.net/ and https://azure.microsoft.com/en-us/pricing/details/virtual-machines/linux/

alexellis commented on cmd/create.go:
```diff
@@ -313,7 +326,23 @@ func createHost(provider, name, region, zone, projectID, userData, inletsPort st
+	} else if provider == "azure" {
+		return &provision.BasicHost{
+			Name:     name,
+			OS:       "Additional.imageOffer",
```
Add a link showing where this is defined in the docs please.

alexellis commented on cmd/create.go:
```diff
@@ -313,7 +326,23 @@ func createHost(provider, name, region, zone, projectID, userData, inletsPort st
+				"pro":            fmt.Sprint(pro),
+				"imagePublisher": "Canonical",
+				"imageOffer":     "UbuntuServer",
+				"imageSku":       "18.04-LTS",
```
We used 16.04 elsewhere as it's smaller and faster. Is that also an option?

alexellis commented on pkg/provision/azure.go:
```diff
@@ -0,0 +1,382 @@
+	}
+
+	log.Printf("Waiting for deployment completion...")
+	err = deploymentFuture.Future.WaitForCompletionRef(p.ctx, deploymentsClient.BaseClient.Client)
```
We need to use non blocking code please.

alexellis commented on pkg/provision/azure.go:
```diff
@@ -0,0 +1,382 @@
+		},
+	)
+	if err != nil {
+		log.Fatalf("failed to create deployment: %v", err)
```
Don't fatal please. Return the error

alexellis commented on pkg/provision/azure.go:
```diff
@@ -0,0 +1,382 @@
+	log.Printf("Group created %s", *group.Name)
+
+	deployment, err := createDeployment(p, host)
+	log.Printf("SSH User: inletsuser / '%s'", host.Additional["adminPassword"])
```
Why is this outputted? We don't use it and it's not strictly relevant.

alexellis commented on pkg/provision/azure.go:
```diff
@@ -0,0 +1,382 @@
+	deployment, err := createDeployment(p, host)
+	log.Printf("SSH User: inletsuser / '%s'", host.Additional["adminPassword"])
+	if err != nil {
+		log.Fatalf("failed to create group: %v", err)
```
Don't fatal. Return the error

alexellis commented:
Please could you work on the feedback and ping me when ready?

alexellis commented on cmd/create.go:
```diff
@@ -313,7 +326,26 @@ func createHost(provider, name, region, zone, projectID, userData, inletsPort st
+			},
+		}, nil
+	} else if provider == "azure" {
+		// Ubuntu images can be found here https://docs.microsoft.com/en-us/azure/virtual-machines/linux/cli-ps-findimage#list-popular-images
```
Thank you for this note

alexellis commented on pkg/provision/azure.go:
```diff
@@ -0,0 +1,375 @@
+		deploymentStatus = ActiveStatus
+	} else {
+		deploymentStatus = *deployment.Properties.ProvisioningState
+		if deploymentStatus == "Running" {
```
Use a const with a name i.e. `AzureActiveStatus` or similar.

alexellis commented on pkg/provision/azure.go:
```diff
@@ -0,0 +1,375 @@
+			deploymentStatus = "deploying"
+		}
+	}
+	IP := "Creating"
```
The IP should never be anything but an IP or empty string, please change?

alexellis commented on pkg/provision/azure.go:
```diff
@@ -0,0 +1,375 @@
+	if err != nil {
+		return err
+	}
+	log.Printf("Waiting for deletion completion (~10 mins)")
```
Please do not wait, this is counter to the way the provisioning package must work, it must return as soon as possible.

alexellis commented on pkg/provision/azure.go:
```diff
@@ -0,0 +1,375 @@
+	}
+}
+func getTemplateResourceNetworkInterface() map[string]interface{} {
+	return map[string]interface{}{
```
This is so verbose compared to other APIs, is there not a way to template or generate this?

Please can you add a HTTP link to the doc where you found it / built it?

zechenbit replied:
There is another way, that we can put all these template and parameters into 2 JSON files. But the JSON files will be even more verbose because in JSON we cannot use go function to generate code with similar struct such as getSecurityRule in this PR. We also need to bundle the JSON files alone with the code when building binary.

This is the link regarding the JSON I'm talking about: https://github.com/Azure-Samples/azure-sdk-for-go-samples/tree/master/quickstarts/deploy-vm

So basically they are doing this:
```
func createDeployment() (deployment resources.DeploymentExtended, err error) {
	template, err := readJSON(templateFile)
	if err != nil {
		return
	}
	params, err := readJSON(parametersFile)
```

This was actually what I did in the first version, and then I figured it may be better to convert all the JSON into go struct / functions and I did it.

alexellis replied:
We could keep a Go literal string instead too, that might be easier for maintenance. I don't want to read a file from disk for this.

zechenbit replied:
If we use literal string, that string will be very large text. And it would add 100+ lines of code in this azure.go because of no go code to generate duplicate/similar lines.

Another con of using literal string is that it actually makes the code logic a bit more complex. For example, right now I can just do:
```
"publisher": host.Additional["imagePublisher"],
```

With literal string, I will need to do something like
```
"publisher": "[parameter('imagePublisher')]",
```
and then add a new Azure parameter `imagePublisher` in `func getTemplateParameterDefinition` and then assign the value in the func `getParameter()`.

Basically from 1 place to 3 places to add a value.

alexellis commented:
Hi @zechenbit please can you check the CI logs and report back on what went wrong? Fix if possible?

Let me know if you need help with that.

alexellis commented:
This PR requires a rebase (please do not do a merge commit)

You may use for instance:

```
git pull origin master --rebase
```

zechenbit commented:
@alexellis The CI is fixed. The failure was due to the go dep to go modules change recently made in master. I fixed that by doing a rebase on the azure branch as you suggested.

I also added the unit tests. And also left a comment regarding using string literals for Azure template.

Please review again at your convenience.

alexellis commented:
Much appreciated. I get back from vacation on Monday. Looking forward to this feature.

alexellis commented:
@zechenbit please can you do a rebase on master? Whatever you do, do not use merge or git pull. `git pull origin master --rebase` may work.

alexellis commented on README.md:
```diff
@@ -172,6 +172,25 @@ inletsctl kfwd --if 192.168.0.14 --from openfaas-figlet:8080
 Then access the service via `http://127.0.0.1:8080`.
 
+### Example usage with Azure
+
```
Please link to how to install the `az` CLI

alexellis commented on cmd/create.go:
```diff
@@ -146,7 +147,14 @@ func runCreate(cmd *cobra.Command, _ []string) error {
+	var accessTokenFile string
+	if provider == "azure" {
+		subscriptionID, _ = cmd.Flags().GetString("subscription-id")
+		accessTokenFile, _ = cmd.Flags().GetString("access-token-file")
```
What if the value is being passed via `--access-token $(cat file.txt)`?

zechenbit replied:
Simply passing in the file contents won't work, because Azure Go SDK expects a file.

A possible way is to use `auth.NewAuthorizerFromEnvironment()`. This way we can allow the user uses 3 parameters and then we assign to the corresponding environment variables below:
```
    - `AZURE_TENANT_ID`: Specifies the Tenant to which to authenticate.
    - `AZURE_CLIENT_ID`: Specifies the app client ID to use.
    - `AZURE_CLIENT_SECRET`: Specifies the app secret to use.
```

Above 3 parameters can be found in the access token file, here is an example:
```
{
  "clientId": "<uuid>"
  "clientSecret": "<uuid>",
  "subscriptionId": "<uuid>",
  "tenantId": "<uuid>",
  "activeDirectoryEndpointUrl": "https://login.microsoftonline.com",
  "resourceManagerEndpointUrl": "https://management.azure.com/",
  "activeDirectoryGraphResourceId": "https://graph.windows.net/",
  "sqlManagementEndpointUrl": "https://management.core.windows.net:8443/",
  "galleryEndpointUrl": "https://gallery.azure.com/",
  "managementEndpointUrl": "https://management.core.windows.net/"
}
```

We may also need to handle the endpoint URL. Azure has different endpoints/base URL: global endpoint as showed above, US Gov, China, etc.

More details:
https://github.com/Azure/azure-sdk-for-go#more-authentication-details

alexellis commented on pkg/provision/azure.go:
```diff
@@ -0,0 +1,368 @@
+	p.resourceGroupName = "inlets-" + host.Name
+	p.deploymentName = "inlets-deploy-" + uuid.New().String()
+
+	log.Printf("Creating resource group %s", p.resourceGroupName)
```
Can you add a `Debug` true/false to the constructor for the provisioner? So that we can turn these messages off when needed.

zechenbit replied:
Currently there seems no debug mode on the top level and no command line flag to switch on/off. Introducing debug mode would need to modify the interface, hence will need to modify all the provisioners. 

Or we introduce some kind of global setting variable that is accessible from everywhere? (Maybe not a good idea though)

alexellis commented on pkg/provision/azure.go:
```diff
@@ -0,0 +1,368 @@
+			Location: to.StringPtr(host.Region)})
+}
+
+func getSecurityRule(name string, priority int, protocol, destPortRange string) map[string]interface{} {
```
For inlets PRO there should be allow `*` i.e. all ports.

alexellis commented:
(The rebase is required before merge)

zechenbit commented:
@alexellis Any idea why there are other commits from master go into this PR after I did the rebase? And one of them has no "signed-off" :https://github.com/inlets/inletsctl/pull/43/commits/99bd3be0e3a10ecd8c345eb33daec6ab8d2d93a3

zechenbit commented:
The `pkg/provision/README.md` change was already in the PR, so no need to create another PR.

Regarding the Azure testing credits, a new sign up will get some credit so after the credit card is verified in the account. The credit will expire after a month or so.
